examples/go/tagging/handler/handler.go

```diff
@@ -31,140 +32,145 @@ type vcConfig struct {
-	secret, err := toml.LoadFile(secretPath)
+// vsConnect connects to vSphere govmomi API using information from vcconfig.toml.
+func vsConnect() error {
+	var (
+		err        error
+		configPath = "/var/openfaas/secrets/vcconfig"
+	)
+
+	// Retrieve vCenter and tag info from the vcconfig file.
+	log.Println("load vcconfig.toml")
+	cfg, err = loadTomlCfg(configPath)
 	if err != nil {
-		return nil, err
+		return fmt.Errorf("vcconfig.toml loading failed: %w", err)
 	}
 
-	cfg := vcConfig{}
-	secret.Unmarshal(&cfg)
+	u := url.URL{
+		Scheme: "https",
+		Host:   cfg.VCenter.Server,
+		Path:   "sdk",
+	}
+	u.User = url.UserPassword(cfg.VCenter.User, cfg.VCenter.Password)
 
-	return cfg, nil
-}
+	// Currently, VEBA does not support signed certificates.
+	insecure := true
 
-// Clients needed to access vSphere APIs
-type Clients struct {
-	Govmomi *govmomi.Client
-	Rest    *rest.Client
+	log.Println("connect to vSphere")
+	client, err = vsClient.New(ctx, u, insecure)
+	if err != nil {
+		return fmt.Errorf("connection to vSphere API failed: %w", err)
+	}
+
+	return nil
 }
 
-// Connect saves connection information to clients.
-func connect(ctx context.Context, vca vCenterAuth) *Clients {
-	var c Clients
+func loadTomlCfg(path string) (vcConfig, error) {
+	cfg := vcConfig{}
 
-	u := url.URL{
-		Scheme: "https",
-		Host:   vca.Server,
-		Path:   "sdk",
+	secret, err := toml.LoadFile(path)
+	if err != nil {
+		return cfg, fmt.Errorf("unable to load vcconfig.toml: %w", err)
 	}
 
-	govc, err := govmomi.NewClient(ctx, &u, true)
+	err = secret.Unmarshal(&cfg)
 	if err != nil {
-		log.Println(err)
+		return cfg, fmt.Errorf("unable to unmarshal vcconfig.toml: %w", err)
 	}
-	c.Govmomi = govc
 
-	u.User = url.UserPassword(vca.User, vca.Password)
+	return cfg, nil
+}
 
-	err = c.Govmomi.Login(ctx, u.User)
+func parseEventMoRef(req []byte) (types.ManagedObjectReference, error) {
+	var event incoming
+	var moRef types.ManagedObjectReference
+
+	err := json.Unmarshal(req, &event)
 	if err != nil {
-		log.Print(err)
+		return moRef, fmt.Errorf("parsing of request failed: %w", err)
 	}
 
-	c.Rest = rest.NewClient(c.Govmomi.Client)
-	c.Rest.Login(ctx, u.User)
+	if event.Data.Vm == nil || event.Data.Vm.Vm.Value == "" {
+		return moRef, errors.New("empty managed reference object")
+	}
+
+	// Fill information in the request into a govmomi type.
+	moRef.Type = event.Data.Vm.Vm.Type
+	moRef.Value = event.Data.Vm.Vm.Value
 
-	return &c
+	return moRef, nil
 }
 
-// Tag adds an existing tag to a VirtualMachine.
-func (clients *Clients) tagVM(ctx context.Context, vm types.ManagedObjectReference, tagID string) {
-	// Get the tag manager which does the tagging.
-	m := tags.NewManager(clients.Rest)
+func handleSignal(ctx context.Context, sigCh <-chan os.Signal) {
+	s := <-sigCh
+	log.Printf("got signal: %v,\nlog out of vSphere", s)
 
-	// Attach tag to VM.
-	err := m.AttachTag(ctx, tagID, vm)
+	err := client.Logout(ctx)
 	if err != nil {
```

Comment:
I'm thinking maybe getting rid of `Logout()` completely because the os signal doesn't transfer from `vmware-even-router` app to the OpenFaaS `handler.go`. And even if it did, all it would do is call `Logout()` which also looks unnecessary given that context shuts down `handler.go`.

What do you think?

Reply:
Your fn sees the OS signal (different process, different signals/handlers). So it's good practice to keep (graceful) shutdown logic.

Reply:
Signal handling has been fixed.